crates/core/tedge_agent/src/file_transfer_server/http_rest.rs

```diff
@@ -1,144 +1,51 @@
+use super::entity_store::entity_store_router;
+use super::file_transfer::file_transfer_router;
+use crate::entity_manager::server::EntityStoreRequest;
+use crate::entity_manager::server::EntityStoreResponse;
 use crate::file_transfer_server::error::FileTransferError;
-use anyhow::anyhow;
-use anyhow::Context;
-use axum::body::StreamBody;
-use axum::routing::get;
 use axum::Router;
-use camino::Utf8Path;
 use camino::Utf8PathBuf;
 use futures::future::FutureExt;
-use hyper::Body;
-use hyper::Request;
-use hyper::StatusCode;
 use rustls::ServerConfig;
 use std::future::Future;
-use std::io::ErrorKind;
-use tedge_actors::futures::StreamExt;
-use tedge_utils::paths::create_directories;
-use tokio::fs::File;
+use tedge_actors::ClientMessageBox;
+use tedge_actors::LoggingSender;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_mqtt_ext::MqttMessage;
 use tokio::io;
-use tokio::io::AsyncBufReadExt;
-use tokio::io::AsyncWriteExt;
-use tokio::io::BufReader;
 use tokio::net::TcpListener;
-use tokio_util::io::ReaderStream;
 
-use super::error::FileTransferRequestError as Error;
-use super::request_files::FileTransferDir;
-use super::request_files::FileTransferPath;
-use super::request_files::RequestPath;
-
-async fn upload_file(
-    path: FileTransferPath,
-    mut request: Request<Body>,
-) -> Result<StatusCode, Error> {
-    fn internal_error(source: impl Into<anyhow::Error>, path: RequestPath) -> Error {
-        Error::Upload {
-            source: source.into(),
-            path,
-        }
-    }
-
-    if let Some(directory) = path.full.parent() {
-        if let Err(err) = create_directories(directory) {
-            return Err(internal_error(err, path.request));
-        }
-
-        match stream_request_body_to_path(&path.full, request.body_mut()).await {
-            Ok(()) => Ok(StatusCode::CREATED),
-            Err(err) if source_err_is_is_a_directory(&err, &path.full) => {
-                Err(Error::CannotUploadDirectory { path: path.request })
-            }
-            Err(err) => Err(internal_error(err, path.request)),
-        }
-    } else {
-        Err(internal_error(
-            anyhow!("cannot retrieve directory name for {}", path.full),
-            path.request,
-        ))
-    }
-}
-
-fn source_err_is_is_a_directory(error: &anyhow::Error, path: &Utf8Path) -> bool {
-    error
-        .downcast_ref()
-        .map(|e| err_is_is_a_directory(e, path))
-        .unwrap_or(false)
+#[derive(Clone)]
+pub(crate) struct AgentState {
+    pub(crate) file_transfer_dir: Utf8PathBuf,
+    pub(crate) entity_store_handle: ClientMessageBox<EntityStoreRequest, EntityStoreResponse>,
+    pub(crate) mqtt_schema: MqttSchema,
+    pub(crate) mqtt_publisher: LoggingSender<MqttMessage>,
```

Comment:
I would delegate to `EntityStoreServer` the responsibility to publish over MQTT the updated entity metadata.

-  `EntityStoreServer`  is already MQTT dependent and publishing entity metadata.
- This HTTP handler would be free of any MQTT dependencies.

Reply:
Resolved by 1a85933b